### ISSUE TITLE
Add SympyProvider class

### ISSUE THREAD

**Issue by diego-plan9** [QISKit/qiskit-addon-sympy#9]
Since this package is becoming kind of the canonical example of a third-party package to be used with `qiskit-core`, it would be awesome if we could also include a fully fledged `Provider` along with it, which would basically allow users to get an instance of the backends by name, as in:

```
get_backend('local_statevector_simulator_sympy')
```

Can we add a `qiskit_addon_sympy/sympyprovider.py` with a `SympyProvider` class (also imported by `__init__.py`)? Ideally it would be a very trimmed down version of [`LocalProvider`](https://github.com/QISKit/qiskit-core/blob/master/qiskit/backends/local/localprovider.py), with no autodiscovery or fancy features.

**Comment by jaygambetta** [comment]
Yes I think this is a must so we can show how to have a 3rd party provider.

**Comment by yaelbh** [comment]
Eric remarks:

Currently adding a backend class means subclassing BaseProvider, BaseBackend, and BaseJob. For local simulators, I suspect people could just use LocalJob as their implementation but I'm not sure that should be described as the standard. To avoid confusion I wonder if the name should change from "local" to "core" for backends that come with qiskit-core?).
 
Implementing the provider would allow wrapper functions to work for the third party classes so e.g. qiskit.available_backends() shows the new backends. I think _wrapper.register may need to be modified to accommodate this. Alternatively we could add another function which calls _wrapper._DEFAULT_PROVIDER.add_provider(new_provider).

**Comment by ajavadia** [comment]
I agree with Erick. We need a SympyProvider class, otherwise the sympy simulator will not show up in `available_backends()`.

I would say let's modify the `register()` function of qiskit, instead of adding another method. We need to make register less tied to ibmq anyway, and this is a good opportunity. Removing `hub`, `group`, `project` is already underway. I think it would be nice to have something like:
To register with ibmq:
`register(token, url, provider_name='ibmq')`
To register with sympy:
`register(provider_name='sympy')`

**Comment by jaygambetta** [comment]
you beat me to it. I also think the examples needs to have two case. showing using register and the provider :-)

**Comment by yaelbh** [comment]
I see two ways to implement @ajavadia's suggestion:

1. Modify `_wrapper.register`:
   ```
   if provider_name == 'ibmq':
       provider = IBMQProvider(token, url,
                               hub, group, project, proxies, verify)
       _DEFAULT_PROVIDER.add_provider(provider)
   elif provider_name == 'sympy':
       try:
            from qiskit_addon_sympy import SympyProvider
            _DEFAULT_PROVIDER.add_provider(SympyProvider())
       except ImportError:
	    ... some error message to notify that the sympy provider is not accessible,
            or simply delete the try-except and pass to the user the error message 
            produced by the failing import ...
   else:
       raise QISKitError('provider name %s is not recognized' % provider_name)
   ```

2. Create `_sympywrapper.py`, therein write a new `register` function:
   ```
   if provider_name == 'ibmq':
       _DEFAULT_PROVIDER.add_provider(SympyProvider())
   else:
       raise QISKitError('provider name %s is not recognized' % provider_name)
   ```

   The user will call `qiskit_addon_sympy.register('sympy')`

Option 1 has the disadvantage that qiskit-core needs to be aware of the third-parties and refer to them.
Option 2 has the disadvantage that the user will have to know to call to the correct `register` function.
Overall I prefer Option 2. Of course I'll be happy to overcome its disadvantage, if it is possible.

**Comment by diego-plan9** [comment]
I think this is an issue that we need to solve at `qiskit-core` level - would it be possible to keep this issue just for the implementation of the `SympyProvider`, assuming that would for the moment be enough, and resolve the registration in `core`?

**Comment by yaelbh** [comment]
In PR #19 I added SympyProvider but did not do anything with register. In the example file it looks like this:
   ```
    # The following lines will be replaced by a usage of register
    provider = SympyProvider()
    _DEFAULT_PROVIDER.add_provider(provider)
   ```

If I implement Option 2 above, it will look like this:
   ```
    sympywrapper.register('sympy')
   ```

**Comment by jaygambetta** [comment]
What you have done is correct. 

We will make register_external(provider) in the core and  it will be all good.

**Comment by ajavadia** [comment]
This is resolved now via #19 and #23.
The new registration method is on qiskit master now, and will be released in v0.5.5 shortly.
Getting qiskit to see the sympy backends is now like this:

First, pip install the `qiskit-addon-sympy` repository. Then:

```python
from qiskit import *
from qiskit_addon_sympy import SympyProvider
register(provider_class=SympyProvider)
available_backends()
```
yields:
```
['local_qasm_simulator',
 'local_statevector_simulator',
 'local_unitary_simulator',
 'sympy_statevector_simulator',
 'sympy_unitary_simulator']
```

```python
q = QuantumRegister(2)
c = ClassicalRegister(2)
circ = QuantumCircuit(q, c)
circ.h(q[0])
circ.cx(q[0], q[1])
result = execute(circ, 'sympy_statevector_simulator').result()
print(result.get_statevector())
```
yields:
```
[sqrt(2)/2 0 0 sqrt(2)/2]
```
